### PR TITLE
Driver: New hierarchy 

### PR DESCRIPTION
**Pull request by panglesd**
Based on #1233.

Implement the new hierarchy in the driver. This time, properly split in commits!

### CODE REVIEW

File: src/driver/odoc_driver.ml [hunk 1]
```diff
@@ -614,6 +614,8 @@ let run libs verbose packages_dir odoc_dir odocl_dir html_dir stats nb_workers
     if voodoo then
       match Util.StringMap.to_list all with
       | [ (_, p) ] ->
+          (* TODO: check whether the [doc] layer should be removed (due to new
```
Comment (jonludlam) [inline src/driver/odoc_driver.ml:617]: This should be fine.

**Comment by jonludlam** [comment]
This looks good, except that there's some code in the voodoo module to determine libraries based on their paths which we need to figure out: https://github.com/ocaml/odoc/blob/master/src/driver/voodoo.ml#L277-L280

**Comment by jonludlam** [comment]
I'm working on a fix for voodoo mode.

**Comment by panglesd** [comment]
Then, I pushed only the fix regarding your other comment.

Is there anything more to do than removing the "lib" intermediate directory in your linked file? I'll wait and see your fix!

**Comment by jonludlam** [comment]
Seems voodoo mode is broken for more reasons, let's just merge this and I'll fix it in another PR.

**Pull request merged**